### PR TITLE
Split ScreenGrabTest into multiple tests. (#1470)

### PR DESCRIPTION
**Pull request by pocmo**
This has two advantages:
* If one of the test fails then the others can still run and we can get
  screenshots for them. Currently if ScreenGrabTest fails then all
  following screenshots are missing.
* Independent tests are more stable and easier to tweak.

### CODE REVIEW

**Comment by pocmo** [comment]
Interestingly I see uiautomator sometimes randomly failing on the emulator locally (API 21). The UI is clearly showing up on the emulator but for some reason uiautomator can't see it. :(

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ErrorPagesScreenshots.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.activity.screenshots;
+
+import android.os.Build;
+import android.support.test.espresso.Espresso;
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ErrorPagesScreenshots.java]: \[Checkstyle\] ERROR: Unused import \- android\.support\.test\.espresso\.Espresso\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ErrorPagesScreenshots.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import static android.support.test.espresso.assertion.ViewAssertions.matches;
+import static android.support.test.espresso.matcher.ViewMatchers.hasFocus;
+import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.web.sugar.Web.onWebView;
+import static android.support.test.espresso.web.webdriver.DriverAtoms.findElement;
+import static android.support.test.espresso.web.webdriver.DriverAtoms.webClick;
+import static android.support.test.espresso.web.webdriver.DriverAtoms.webScrollIntoView;
+import static junit.framework.Assert.assertTrue;
+import static org.mozilla.focus.activity.TestHelper.browserURLbar;
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ErrorPagesScreenshots.java]: \[Checkstyle\] ERROR: Unused import \- org\.mozilla\.focus\.activity\.TestHelper\.browserURLbar\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ErrorPagesScreenshots.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            assertTrue(TestHelper.progressBar.waitUntilGone(waitingTime));
+
+            // Android O has an issue with using Locator.ID
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                UiObject tryAgainBtn = device.findObject(new UiSelector()
+                        .descriptionContains(getString(R.string.errorpage_refresh))
+                        .clickable(true));
+                assertTrue(tryAgainBtn.waitForExists(waitingTime));
+            } else {
+                onWebView()
```
Comment (npark-mozilla) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ErrorPagesScreenshots.java]: So when I run a suite locally on the sim, there are no issues, but when I run this test alone, I always run into an error here: 

Caused by: java.util.concurrent.TimeoutException: Waited 10 seconds for android.support.test.espresso.core.internal.deps.guava.util.concurrent.AbstractTransformFuture$TransformFuture@2c24f97[status=PENDING, info=[inputFuture=[android.support.test.espresso.core.internal.deps.guava.util.concurrent.SettableFuture@6c61584[status=PENDING]], function=[android.support.test.espresso.web.action.AtomAction$3@4fd916d]]]

This seems like some internal espresso process is blocked by something.  This is also why I replaced this part with the uiautomator PR. (https://github.com/mozilla-mobile/focus-android/issues/1415)

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/BrowserScreenScreenshots.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        onView(withId(R.id.url_edit))
+                .check(matches(isDisplayed()))
+                .check(matches(hasFocus()))
+                .perform(click(), replaceText(webServer.url("/").toString()), pressImeActionButton());
+
+        device.findObject(new UiSelector()
+                .resourceId("org.mozilla.focus.debug:id/webview")
+                .enabled(true))
+                .waitForExists(waitingTime);
+
```
Comment (npark-mozilla) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/BrowserScreenScreenshots.java]: let's also add check of progressBar to disappear. Sometimes webview appears, but it's still loading.  (esp. on slower sims)

public static UiObject progressBar = mDevice.findObject(new UiSelector()
            .resourceId("org.mozilla.focus.debug:id/progress")
            .enabled(true));

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ErrorPagesScreenshots.java (hunk 3 above)
Comment (pocmo) [reply app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ErrorPagesScreenshots.java]: On what configuration are you running the tests?

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/NotificationScreenshots.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                .enabled(true));
+
+        device.openNotification();
+
+        try {
+            if (!openAction.waitForExists(waitingTime)) {
+                // The notification is not expanded. Let's expand it now.
+                device.findObject(new UiSelector()
+                        .text(getString(R.string.app_name)))
+                        .swipeDown(5);
```
Comment (npark-mozilla) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/NotificationScreenshots.java]: I noticed low step number of .swipeDown() can cause unpredictable results, try experimenting with double digit numbers (for me 50 seems pretty safe)

Comment (pocmo) [reply app/src/androidTest/java/org/mozilla/focus/activity/screenshots/NotificationScreenshots.java]: Oh, I forgot to update this PR. Locally I was using 50. On some emulators it seems to trigger a long-press which is super awkward.

**Comment by pocmo** [comment]
> Also, just want to double check, are all onView() enclosed checks come with waiting for a few seconds? (they probably are, but if not, we should make them wait)

This shouldn't be needed in Espresso world: Espresso will wait for all UI thread operations to be completed before moving on (idle state). And if this is not enough then there are ways to add additional state checks that need to be done before the device is considered idle (e.g. I wrote SessionLoadedIdlingResource which - when activated for the test - will wait for the current website to be loaded completely before moving on).

However in some tests we are mixing uiautomator and espresso which might lead to problems. That's one of the reason why I would like to go with espresso-only tests where applicable (only UI in-app).

**Pull request merged**